server.go

```diff
@@ -15,13 +47,101 @@ type Server struct {
 
 	// Handler to invoke. If nil, uses DefaultHandler.
 	Handler Handler
+
+	// Shutdown handling
+	// --
+	lock         sync.RWMutex
+	started      bool
+	wg           sync.WaitGroup            // active workers (servers)
+	tcpListeners map[net.Listener]struct{} // track active TCP listeners
+	udpListeners map[*net.UDPConn]struct{} // track active UDP listeners
+	tcpConns     map[net.Conn]struct{}     // track active connections
+}
+
+// type check
+var _ ServerDNSCrypt = &Server{}
+
+// Shutdown - tries to gracefully shutdown the server. It waits until all
+// connections are processed and only after that it leaves the method.
+// If context deadline is specified, it will exit earlier.
+func (s *Server) Shutdown(ctx context.Context) error {
+	log.Info("Shutting down the DNSCrypt server")
+
+	s.lock.Lock()
+	if !s.started {
+		s.lock.Unlock()
+		log.Info("Server is not started")
+		return ErrServerNotStarted
+	}
+
+	s.started = false
+
+	// Unblock reads for all active tcpConns
+	for conn := range s.tcpConns {
+		_ = conn.SetReadDeadline(time.Unix(1, 0))
+	}
+
+	// Unblock reads for all active TCP listeners
+	for l := range s.tcpListeners {
+		switch v := l.(type) {
+		case *net.TCPListener:
+			_ = v.SetDeadline(time.Unix(1, 0))
+		}
+	}
+
+	// Unblock reads for all active UDP listeners
+	for l := range s.udpListeners {
+		_ = l.SetReadDeadline(time.Unix(1, 0))
+	}
+	s.lock.Unlock()
+
+	// Using this channel to wait until all goroutines finish their work
+	closed := make(chan struct{})
+	go func() {
+		s.wg.Wait()
+		log.Info("Serve goroutines finished their work")
+		close(closed)
+	}()
+
+	// Wait for either all goroutines finish their work
+	// Or for the context deadline
+	var ctxErr error
+	select {
+	case <-closed:
+		log.Info("DNSCrypt server has been stopped")
+	case <-ctx.Done():
+		log.Info("DNSCrypt server shutdown has timed out")
+		ctxErr = ctx.Err()
+	}
+
+	return ctxErr
+}
+
+// init - initializes (lazily) Server properties on startup
+// this method is called from Server.ServeTCP and Server.ServeUDP
+func (s *Server) init() {
```

Comment:
Is there a constructor?  If there is, why not do that there?  Lazy initialisation generally becomes a source of major pains later.

Reply:
I wouldn't say that that is a good API to replicate, to be honest.  If anything, the amount of people who forget to set timeouts on their `http.Server` proves that it is a bad idea, and this server seems to only have three exported parameters.  I already see the `if s.started` and `if s.isStarted()` conditionals popping up here and there, and that just begs for lots of issues down the line.